### PR TITLE
use map-reduce method for all priorities method

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: kadisi <iamkadisi@163.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
use map-reduce method for all priorities method
```

### CODE REVIEW

**Comment by jayunit100** [comment]
Is there an upstream issue for implementing this?  i.e. lets make one comparing the mapreduce impl suggested here to the current parallelization that is done...

**Comment by resouer** [comment]
@kadisi Thanks for contribution, while this is not a small change. 

So as @jayunit100 mentioned, could you please create a issue to clarify the problem this PR trying solve? Describe the basic idea you proposed in this patch? It would be very helpful.

**Comment by kadisi** [comment]
@resouer thank you, i opend an issue: [](https://github.com/kubernetes/kubernetes/issues/63127)

**Comment by kadisi** [comment]
@wgliang thank you !

**Comment by wgliang** [comment]
cc @bsalamat @k82cn Do we need to modify the existing method of dealing with priorities. In addition, I think that you should evaluate performance changes. Can you supplement your performance test results? @kadisi

**Comment by wgliang** [comment]
/hold

hold before we have an exact resolution in #63127, and @kadisi you can supplement your test results and instructions.

**Comment by kadisi** [comment]
@wgliang  ，how to supplement  my performance test results ?

**Comment by wgliang** [comment]
@kadisi Yes, you need to compare the results of performance testing (real-world testing or integration testing) to prove that your improvement is at least higher than current performance, or at least not bad. Then we can consider accepting this PR.

**Comment by kadisi** [comment]
@wgliang the testing performance is 
```
I0509 06:34:42.941837    2785 factory.go:1066] Created equivalence class cache
0s	rate: 448	total: 448 (qps frequency: map[448:1])
1s	rate: 323	total: 771 (qps frequency: map[323:1 448:1])
2s	rate: 331	total: 1102 (qps frequency: map[448:1 323:1 331:1])
3s	rate: 306	total: 1408 (qps frequency: map[448:1 323:1 331:1 306:1])
4s	rate: 320	total: 1728 (qps frequency: map[331:1 306:1 320:1 448:1 323:1])
5s	rate: 307	total: 2035 (qps frequency: map[306:1 320:1 307:1 448:1 323:1 331:1])
6s	rate: 316	total: 2351 (qps frequency: map[316:1 448:1 323:1 331:1 306:1 320:1 307:1])
7s	rate: 334	total: 2685 (qps frequency: map[448:1 323:1 331:1 306:1 320:1 307:1 316:1 334:1])
Scheduled 3000 Pods in 8 seconds (375 per second on average). min QPS was 306
I0509 06:34:57.484268    2785 util.go:83] destroying scheduler
I0509 06:34:57.484286    2785 util.go:87] destroyed scheduler
I0509 06:34:57.484290    2785 util.go:51] destroying API server
I0509 06:34:57.486599    2785 util.go:53] destroyed API server
Minimal observed throughput for 3k pod test: 306
PASS
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	18.851s
```

[https://github.com/kubernetes/kubernetes/issues/63127](url)

**Comment by wgliang** [comment]
/cc @bsalamat @k82cn
PTAL

File: pkg/scheduler/algorithm/priorities/metadata.go [hunk 1]
```diff
@@ -17,56 +17,268 @@ limitations under the License.
...
 }
 
 // PriorityMetadata is a PriorityMetadataProducer.  Node info can be nil.
-func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo) interface{} {
+func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodes []*v1.Node) interface{} {
 	// If we cannot compute metadata, just return nil
 	if pod == nil {
 		return nil
 	}
+	hpMap, _ := pmf.calculateInterPodAffinityPriority(pod, nodeNameToInfo, nodes)
```
[the remark this answers is not in the document]
Comment (kadisi) [reply pkg/scheduler/algorithm/priorities/metadata.go]: sorry， It's not very precise.

File: pkg/scheduler/algorithm/priorities/metadata_test.go [hunk 2]
```diff
@@ -26,10 +27,610 @@ import (
...
+				{ObjectMeta: metav1.ObjectMeta{Name: "machine2", Labels: labelRgIndia}},
+				{ObjectMeta: metav1.ObjectMeta{Name: "machine3", Labels: labelAzAz1}},
+			},
+			hardPodAffinityWeight: 0,
+			expectedList:          map[string]schedulerapi.HostPriority{"machine1": {Host: "machine1", Score: 0}, "machine2": {Host: "machine2", Score: 0}, "machine3": {Host: "machine3", Score: 0}},
+			test:                  "Hard Pod Affinity symmetry: hard pod affinity symmetry is closed(weights 0), then nodes that match the hard pod affinity symmetry rules, get same score with those not match",
+		},
+	}
+	for _, test := range tests {
+		nodeNameToInfo := schedulercache.CreateNodeNameToInfoMap(test.pods, test.nodes)
```
[the remark this answers is not in the document]
Comment (kadisi) [reply pkg/scheduler/algorithm/priorities/metadata_test.go]: thank you

File: pkg/scheduler/algorithm/priorities/metadata_test.go [hunk 3]
```diff
@@ -117,13 +718,13 @@ func TestPriorityMetadata(t *testing.T) {
 	}
 	tests := []struct {
 		pod      *v1.Pod
-		name     string
+		test     string
```
[the remark this answers is not in the document]
Comment (kadisi) [reply pkg/scheduler/algorithm/priorities/metadata_test.go]: ok

**Comment by bsalamat** [comment]
@aveshagarwal Could you please take a look at this?

**Comment by aveshagarwal** [comment]
Sure.

**Comment by aveshagarwal** [comment]
@kadisi will review it soon.

**Comment by kadisi** [comment]
@aveshagarwal  thank you

**Comment by kadisi** [comment]
?

**Pull request closed without merging**